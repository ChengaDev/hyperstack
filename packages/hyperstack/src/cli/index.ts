/* eslint-disable @typescript-eslint/no-var-requires */
import yargs from 'yargs'

export const cli = async (root: string, argv: string[]) => {
  return await yargs(argv.slice(2))
    .config({ root })
    .command(require('./cmds/migrate').default)
    .command(require('./cmds/generate').default)
    .command(require('./cmds/seed').default)
    .command(require('./cmds/routes').default)
    .command(require('./cmds/start').default)
    .command(require('./cmds/portal').default)
    .command(require('./cmds/tasks').default)
    .demandCommand()
    .help().argv
}
