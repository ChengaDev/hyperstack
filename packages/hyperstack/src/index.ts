// expose the _right_ surface-area for users

export * from './tie'
export { initializer } from './initializers'
export { seed } from './models'
export { task } from './tasks'
export { isDevelopment, isProduction, context } from './context'
export { cli } from './cli'
export type {
  Context,
  Initializer,
  MigrationsMode,
  WorkersMode,
  Task,
  App,
  Config,
} from './types'

export {
  Schema,
  HyperModel,
  HyperModelError,
  ModelAssertionError,
  hashPassword,
  verifyPassword,
  generateHexToken,
} from '@hyperstackjs/hypermodel'

export {
  ClassErrorMiddleware,
  ClassMiddleware,
  Controller,
  Delete,
  ErrorMiddleware,
  Get,
  HttpResponse,
  HttpResponseBadRequest,
  HttpResponseClientError,
  HttpResponseConflict,
  HttpResponseCreated,
  HttpResponseForbidden,
  HttpResponseInternalServerError,
  HttpResponseMethodNotAllowed,
  HttpResponseMovedPermanently,
  HttpResponseNoContent,
  HttpResponseNotFound,
  HttpResponseNotImplemented,
  HttpResponseOK,
  HttpResponseRedirect,
  HttpResponseRedirection,
  HttpResponseServerError,
  HttpResponseSuccess,
  HttpResponseTooManyRequests,
  HttpResponseUnauthorized,
  Middleware,
  ParseBody,
  Patch,
  Post,
  Put,
  ValidateBody,
  ValidationError,
  authWithJWT,
  err,
  notfound,
  ok,
  permits,
  requires,
  unauthorized,
  Request,
  Response,
  NextFunction,
} from '@hyperstackjs/hypercontroller'

export {
  HyperWorker,
  Mailer,
  queueAs,
  producerOnly,
} from '@hyperstackjs/hyperworker'
