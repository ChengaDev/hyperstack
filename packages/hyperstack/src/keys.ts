export const setters = {
  tie: 'tie',
  initializers: 'initializers',
  context: 'context',
  controllers: 'controllers',
  logger: 'logger',
  models: 'models',
  tasks: 'tasks',
  workers: 'workers',
}

export const consts = {
  appEnvKey: 'HST_ENV',
}
