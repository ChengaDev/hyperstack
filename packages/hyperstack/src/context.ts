import path from 'path'
import createDebug from 'debug'
import L from 'lodash'
import type { Config, ContextStore } from '@hyperstackjs/typings'
import { consts, setters } from './keys'
import type { Context } from './types'
import { printJobCounts, printRoutesV2 } from './utils'

const debug = createDebug('hyperstack:context')

class InMemContext implements Context {
  private __store: ContextStore = {}
  private __config: Config | null = null
  clear = () => {
    this.__store = {}
    this.__config = null
  }

  baseUrl = () => this.__store.controllers!.baseUrl
  logger = () => this.__store.logger
  models = () => this.__store.models!.models
  mailers = () => this.__store.workers!.mailers
  workers = () => this.__store.workers!.workers
  controllers = () => this.__store.controllers!.controllers
  environment = () => this.__store.app!.environmentFile
  helpers = {
    printRoutes: printRoutesV2,
    printJobCounts,
  }

  update = (updater: string, d: Partial<ContextStore>) => {
    debug('[%o] context update: %o', updater, d)
    L.merge(this.__store, d)
  }

  setConfig = (c: Config) => {
    this.__config = c
  }

  store = () => this.__store
  config = () => {
    if (!this.__config) {
      throw new Error('configuration not loaded yet.')
    }
    return this.__config
  }

  initializerProps = <T>(key: string) =>
    L.get(this.__store.initializers?.props || {}, key) as T
}

const context = new InMemContext()

export { context }

export const isTest = () => process.env.NODE_ENV === 'test'
export const isProduction = () => process.env.NODE_ENV === 'production'
export const isDevelopment = () => !isTest() && !isProduction()

export const envMode = () =>
  isTest() ? 'test' : isProduction() ? 'production' : 'development'
export const getEnvironment = () =>
  L.replace(process.env.NODE_ENV || 'development', /[^A-Za-z0-9_-]/g, '')

export const getAppEnvironment = () =>
  process.env[consts.appEnvKey]
    ? L.replace(process.env[consts.appEnvKey]!, /[^A-Za-z0-9_-]/g, '')
    : getEnvironment()

export const resolveAppRoot = (app?: string) =>
  app || path.join(process.cwd(), '/src')

export const tieApp = (context: Context, app?: string) => {
  const appRoot = resolveAppRoot(app)
  context.update(setters.context, {
    app: {
      mode: envMode(),
      environmentFile: getAppEnvironment(),
      root: appRoot,
    },
  })
}
