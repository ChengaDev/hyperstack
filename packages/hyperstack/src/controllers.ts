import path from 'path'
import fs from 'fs'
import type { ServerOptions } from '@hyperstackjs/hypercontroller'
import { createServer } from '@hyperstackjs/hypercontroller'
import type { Context } from './types'
import { setters } from './keys'
import { requireModule } from './utils'

const VAN_HALEN_PORT = 5150
const resolvePort = () => process.env.PORT || VAN_HALEN_PORT
const createApp = ({
  controllers,
  initializers,
  logging,
  staticFolder,
  indexCatchAll,
  forceHttps,
  gzip,
  cookieSecret,
  sendValidationErrors,
}: { controllers: any; initializers: any } & Partial<ServerOptions>) => {
  return createServer({
    controllers,
    opts: {
      logging,
      staticFolder,
      indexCatchAll,
      forceHttps,
      gzip,
      cookieSecret,
      sendValidationErrors,
      beforeMiddleware: (preapp: any) => {
        if (initializers && initializers.beforeMiddleware) {
          initializers.beforeMiddleware.forEach((init: (app: any) => void) =>
            init(preapp)
          )
        }
      },
      afterMiddleware: (postapp: any) => {
        if (initializers && initializers.afterMiddleware) {
          initializers.afterMiddleware.forEach((init: (app: any) => void) =>
            init(postapp)
          )
        }
      },
      beforeControllers: (preapp: any) => {
        if (initializers && initializers.beforeControllers) {
          initializers.beforeControllers.forEach((init: (app: any) => void) =>
            init(preapp)
          )
        }
      },
      afterControllers: (postapp: any) => {
        if (initializers && initializers.afterControllers) {
          initializers.afterControllers.forEach((init: (app: any) => void) =>
            init(postapp)
          )
        }
      },
    },
  })
}

export const getControllerConfig = (context: Context) => {
  const store = context.store()
  const appRoot = store.app!.root
  const logger = store.logger
  const middleware = store.logging!.middleware
  const initializers = store.initializers
  const controllerPath = path.join(appRoot, 'app', 'controllers')

  const {
    serveStatic,
    indexCatchAll,
    forceHttps,
    gzip,
    cookieSecret,
    sendValidationErrors,
    json,
    helmet,
    urlencoded,
    baseUrl: cBaseUrl,
  } = context.config().controllers || {}

  const baseUrl = cBaseUrl || `http://localhost:${resolvePort()}`

  if (!fs.existsSync(controllerPath)) {
    throw new Error(`tie controllers: no app controller in ${controllerPath}`)
  }
  const { AppController } = requireModule(controllerPath)
  if (!AppController) {
    throw new Error(`No 'AppController' export was found in ${controllerPath}`)
  }

  const staticFolder = serveStatic ? path.join(appRoot, 'public') : undefined
  return {
    controllers: AppController.controllers,
    initializers,
    logging: { logger, middleware },
    staticFolder,
    indexCatchAll,
    forceHttps,
    gzip,
    cookieSecret,
    sendValidationErrors,
    helmet,
    json,
    urlencoded,
    baseUrl,
  }
}

export const tieControllers = async (context: Context) => {
  const appConfig = getControllerConfig(context)
  const app = createApp(appConfig)

  context.update(setters.controllers, {
    controllers: {
      numControllers: appConfig.controllers.length,
      baseUrl: appConfig.baseUrl,
      controllers: appConfig.controllers,
      port: resolvePort(),
      app,
      didTieControllers: true,
    },
  })
}
