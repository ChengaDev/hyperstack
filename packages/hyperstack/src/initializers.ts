import path from 'path'
import L from 'lodash'
import createDebug from 'debug'
import { load } from './utils'
import type {
  Context,
  InitializerCreator,
  LifecycleHooks,
  LifecycleMap,
} from './types'
import { setters } from './keys'
const debug = createDebug('hyperstack:initializers')

export const initializer = (fn: InitializerCreator) => {
  return fn
}

export const getInitializers = (appRoot: string): InitializerCreator[] => {
  const initpath = path.join(appRoot, 'config/initializers/*.[tj]s')
  return (load(initpath) as InitializerCreator[]) || []
}

export const tieInitializers = async (context: Context) => {
  const appRoot = context.store().app!.root
  const initializerCreators = getInitializers(appRoot)
  const initializers: LifecycleHooks[] = []
  for (const initializerCreator of initializerCreators) {
    initializers.push(await initializerCreator(context))
  }

  // into initializermap, array initializers per lifecycle hook:
  // [{beforeApp:}, {beforeApp:}] -> { beforeApp: [..], ... afterApp: [..]}
  const inits = L.mergeWith({}, ...initializers, (v1: any, v2: any) =>
    L.isUndefined(v1) ? [v2] : [...v1, v2]
  ) as LifecycleMap

  inits.beforeApp && inits.beforeApp.forEach((init) => init())

  const props = {}
  inits.provideProps &&
    inits.provideProps.forEach((provide) => {
      const p = provide()
      debug('providing props: %o', p)
      L.merge(props, p)
    })

  context.update(setters.initializers, {
    initializers: { initializers: inits, props },
  })
}
