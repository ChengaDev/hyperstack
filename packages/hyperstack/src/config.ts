import path from 'path'
import L from 'lodash'
import createDebug from 'debug'
import type { Config } from '@hyperstackjs/typings'
import { getAppEnvironment } from './context'
import { requireModule } from './utils'
import type { ConfigPaths, Context } from './types'
const debug = createDebug('hyperstack:config')

export const resolveConfig = ({ exactPath, appPath }: ConfigPaths) => {
  if (exactPath) {
    return exactPath
  }
  if (!appPath) {
    throw new Error('app path is empty')
  }
  return path.join(appPath, `config/environments/${getAppEnvironment()}`)
}
export const loadConfig = async (paths: ConfigPaths) => {
  const resolved = resolveConfig(paths)

  const mod = requireModule(resolved)

  // configs are function providing a dict.
  // it's a function because merging modifies the provided dict, and we
  // don't want to fiddle with module cache in case we want live reloads or tests to work
  const config = (await mod()) as Config

  debug('config before overrides is %o', config)
  if (paths.configOverrides) {
    L.merge(config, paths.configOverrides)
  }
  return { config, resolvedConfigPath: resolved }
}

export const tieConfig = async (
  context: Context,
  { exactPath, configOverrides }: ConfigPaths = {}
) => {
  const appPath = context.store().app!.root
  const { config, resolvedConfigPath } = await loadConfig({
    configOverrides,
    exactPath,
    appPath,
  })
  debug('loaded from %o', resolvedConfigPath)
  debug('config is %o', config)
  debug('overrides %o', configOverrides)
  context.setConfig(config)
}
