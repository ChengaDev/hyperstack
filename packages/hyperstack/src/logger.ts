import { createColors } from 'colorette'
import L from 'lodash'
import createDebug from 'debug'
import type P from 'pino'
import pino from 'pino'
import type { AppLoggerOptions, Logger } from '@hyperstackjs/typings'
import pretty from 'pino-pretty'
import createPinoMiddleware from 'express-pino-logger'
import { v4 as uuidv4 } from 'uuid'
import type { Request } from '@hyperstackjs/hypercontroller'
import { setters } from './keys'
import type { Context } from './types'
import { isDevelopment, isProduction, isTest } from './context'
const debug = createDebug('hyperstack:logger')

const devPretty = pretty({
  sync: true,
  colorize: true,
  levelFirst: true,
  translateTime: true,
  singleLine: true,
  ignore: 'pid,hostname',
})

/**
 * Types from Pino
 * @see https://github.com/pinojs/pino/blob/master/pino.d.ts
 */
export type BaseLogger = P.BaseLogger
export type DestinationStream = P.DestinationStream
export type LevelWithSilent = P.LevelWithSilent
export type LoggerOptions = P.LoggerOptions

export const redactionsList: string[] = [
  'access_token',
  'data.access_token',
  'data.*.access_token',
  'data.*.accessToken',
  'accessToken',
  'data.accessToken',
  'DATABASE_URL',
  'data.*.email',
  'data.email',
  'email',
  'event.headers.authorization',
  'data.hashedPassword',
  'data.*.hashedPassword',
  'hashedPassword',
  'host',
  'jwt',
  'data.jwt',
  'data.*.jwt',
  'JWT',
  'data.JWT',
  'data.*.JWT',
  'password',
  'data.password',
  'data.*.password',
  'params',
  'data.salt',
  'data.*.salt',
  'salt',
  'secret',
  'data.secret',
  'data.*.secret',
]

export const logLevel = () => {
  if (typeof process.env.LOG_LEVEL !== 'undefined') {
    return process.env.LOG_LEVEL
  } else if (isProduction()) {
    return 'warn'
  } else if (isTest()) {
    return 'silent'
  } else {
    return 'trace'
  }
}

export const defaultLoggerOptions: Partial<AppLoggerOptions> = {
  level: logLevel(),
  redact: {
    paths: redactionsList,
  },
}

export const createLogger = (loggerOpts: AppLoggerOptions = {}): Logger => {
  const options = L.merge({}, defaultLoggerOptions, loggerOpts)

  debug('Logger Configuration')
  debug(`isProduction: ${isProduction()}`)
  debug(`isDevelopment: ${isDevelopment()}`)
  debug(`isTest: ${isTest()}`)
  debug(`logLevel: ${options.level}`)
  debug(`stream: ${!!options.stream}`)

  const pinoOpts = L.omit(options, ['stream', 'middleware'])
  const colors = createColors({ useColor: !isProduction() })

  let logger = null
  if (isProduction()) {
    logger = pino(
      pinoOpts as LoggerOptions,
      options.stream as DestinationStream
    ) as any as Logger
  } else {
    debug('Logging to dev environment: pretty mode')
    logger = pino(pinoOpts as LoggerOptions, devPretty) as any as Logger
  }

  logger.colors = colors
  return logger
}

export const customRequestProps = (req: Request, _res: Response) => {
  if (req.user) {
    const user = req.user
    // sniff out a reasonable identifier
    const out: any = {}
    ;['id', 'pid', 'email', 'username'].forEach((k) => {
      if (user[k]) {
        out[`user-${k}`] = user[k]
      }
    })
    return out
  }
  return {}
}

export const createMiddleware = (logger: Logger, middlewareConfig?: any) => {
  return createPinoMiddleware({
    logger,
    genReqId: (req) => req.id || uuidv4().toString(),
    customProps: customRequestProps,
    ...(middlewareConfig || {}),
  })
}

export const tieLogger = (context: Context) => {
  const loggerConfig = context.config().logger
  const logger = createLogger(loggerConfig)
  const middlewareConfig = loggerConfig?.middleware
  const middleware = createMiddleware(logger, middlewareConfig)

  const initializers = context.store().initializers!.initializers
  const { logger: theirLogger, middleware: theirMiddleware } =
    initializers.beforeLogger?.reduce(
      (res, init) => {
        const out = init(res)
        return out || res
      },
      { logger, middleware }
    ) || {}

  // user can replace impl. of colors, but if none
  // given, we want ours.
  if (theirLogger && !theirLogger.colors) {
    theirLogger.colors = createColors({ useColor: !isProduction() })
  }

  const selectedLogger = theirLogger || logger
  const selectedMiddleware = theirMiddleware || middleware

  context.update(setters.logger, {
    logging: {
      middleware: selectedMiddleware,
    },
    logger: selectedLogger,
  })
  return logger
}
