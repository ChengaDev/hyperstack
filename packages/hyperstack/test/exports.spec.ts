import * as API from '../src'

describe('exports', () => {
  it('records api surface area', () => {
    expect(Object.keys(API)).toMatchSnapshot()
  })
})
