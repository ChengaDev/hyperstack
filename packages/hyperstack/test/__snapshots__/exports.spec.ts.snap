// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`exports records api surface area 1`] = `
[
  "initializer",
  "seed",
  "task",
  "isDevelopment",
  "isProduction",
  "context",
  "cli",
  "Schema",
  "HyperModel",
  "HyperModelError",
  "ModelAssertionError",
  "hashPassword",
  "verifyPassword",
  "generateHexToken",
  "ClassErrorMiddleware",
  "ClassMiddleware",
  "Controller",
  "Delete",
  "ErrorMiddleware",
  "Get",
  "HttpResponse",
  "HttpResponseBadRequest",
  "HttpResponseClientError",
  "HttpResponseConflict",
  "HttpResponseCreated",
  "HttpResponseForbidden",
  "HttpResponseInternalServerError",
  "HttpResponseMethodNotAllowed",
  "HttpResponseMovedPermanently",
  "HttpResponseNoContent",
  "HttpResponseNotFound",
  "HttpResponseNotImplemented",
  "HttpResponseOK",
  "HttpResponseRedirect",
  "HttpResponseRedirection",
  "HttpResponseServerError",
  "HttpResponseSuccess",
  "HttpResponseTooManyRequests",
  "HttpResponseUnauthorized",
  "Middleware",
  "ParseBody",
  "Patch",
  "Post",
  "Put",
  "ValidateBody",
  "ValidationError",
  "authWithJWT",
  "err",
  "notfound",
  "ok",
  "permits",
  "requires",
  "unauthorized",
  "HyperWorker",
  "Mailer",
  "queueAs",
  "producerOnly",
  "createApp",
  "shutDown",
  "createUseApp",
  "useWorkers",
  "usePortal",
  "useApp",
  "useModels",
  "useControllers",
  "useTasks",
  "useSeed",
  "runApp",
]
`;
